### ISSUE TITLE
Feature Request for a Synchronous way to Draw Raw Pixel data to Canvas

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [flutter/flutter#37180]
## Use case

This feature request is about the [`Canvas` class in `engine/painting.dart`](https://github.com/flutter/engine/blob/master/lib/ui/painting.dart#L3105).

As extensively discussed in #31598, using regular canvas operations to draw pixel data is *extremely slow*.  
This makes the only option we have decoding an image from our raw pixel data, e.g. using [`decodeImageFromPixels`](https://api.flutter.dev/flutter/dart-ui/decodeImageFromPixels.html). This has another problem, [pointed out by @andrewackerman in the mentioned issue](https://github.com/flutter/flutter/issues/31598#issuecomment-511510444), which is that creating an `Image` in Dart is an `async` operations.  
This means that we will have to postpone drawing our raw pixel data to the next frame if we wish to use real-time operations.

## Proposal

[@andrewackerman already wrote a comprehensive proposal in #31598](https://github.com/flutter/flutter/issues/31598#issuecomment-516050204), so I am just going to quote that here:

> [...] it should have a similar signature to `decodImageFromPixels` [sic], so something like:

```dart
class Canvas {
  ...

  void drawPixels(
    Offset c, 
    Uint8List bytes,
    int width,
    int height,
    PixelFormat pixelFormat, {
    int rowBytes,
    int targetWidth,
    int targetHeight,
  }) {
    assert(c != null && bytes != null && width != null && height != null && pixelFormat != null);
    assert(bytes.length == width * height * 4);
    if (rowBytes == null) rowBytes = width * 4;
    if (targetWidth == null && targetHeight == null) {
      targetWidth = width;
      targetHeight = height;
    } else if (targetWidth == null) {
      targetWidth = (targetHeight * (width / height)).toInt();
    } else if (targetHeight == null) {
      targetHeight = (targetWidth * (height / width)).toInt();
    }
    _drawPixels(c, pixelData, width, height, pixelFormat, rowBytes, targetWidth, targetHeight);
  }
}
```

> `c` is the position that the image will be drawn at.
>
> `bytes` is the pixel data that will be drawn to the canvas, represented as byte groups of 4 bytes per pixel. The color channels will be extrapolated based on the value of `pixelFormat`.
>
> `width` and `height` are the dimensions of the image represented by the pixel data. If `width * height * 4` is not equal to `pixelData.length`, an error will be thrown.
>
> `rowBytes` is the number of bytes consumed by each row of pixels in the data buffer. If unspecified, it defaults to width multiplied by the number of bytes per pixel in the provided `format`.
>
> The `targetWidth and `targetHeight` arguments specify the size of the output image, in image pixels. If they are not equal to the intrinsic dimensions of the image, then the image will be draw. If exactly one of these two arguments is specified, then the aspect ratio will be maintained while forcing the image to match the other given dimension. If neither is specified, then the image maintains its real size. (Debatable as to whether these parameters should be supported as they don't really fall into the whole "just draw these pixels" mentality of this method.)**

@andrewackerman assumed that the byte data will always have 32 bits per pixel, which makes sense because the only [`PixelFormat` values](https://api.flutter.dev/flutter/dart-ui/PixelFormat-class.html) that currently exist are `bgra8888` and `rgba8888`, which use 32 bits per pixel.  
The [`decodeImageFromPixels` function](https://api.flutter.dev/flutter/dart-ui/decodeImageFromPixels.html) is open to any `PixelFormat` that might be supported in the future:

> `rowBytes` is the number of bytes consumed by each row of pixels in the data buffer. If unspecified, it defaults to width multiplied by the number of bytes per pixel in the provided `format`.

I think that this approach would make sense to also be used for the `Canvas` method (if it will use `PixelFormat`). Consequently, the assertion for `width * height * 4` should probably be `width * height * bytesPerPixel` and similarly, `bytes` should be pixel data, represented as byte groups of `bytesPerPixel` bytes per pixel.

**Comment by Solido** [comment]
int width, int height & int targetWidth, int targetHeight should be replaced with Size class

**Comment by andrewackerman** [comment]
Ideally, though this signature matches the style of the next closest existing method for this sort of thing ([`decodeImageFromPixels`](https://api.flutter.dev/flutter/dart-ui/decodeImageFromPixels.html)). I could see arguments both for leaving it "low-levelish" using only primitives and for standardizing it with `Size`.

**Comment by Solido** [comment]
I get the idea but this method would not only be used in conjunction with `decodeImageFromPixels ` and it can feel strange to have different pattern that other methods in Canvas ?

`
void drawImageRect (
   Image image,
   Rect src,
   Rect dst,
   Paint paint
)
`

**Comment by creativecreatorormaybenot** [comment]
@Solido Well, the other canvas methods are synchronous. It would just have to be a little faster..

**Comment by Solido** [comment]
Others methods are native. Not sure that what we're asking for is available is the same way else 
`drawPoints` would be enough

**Comment by creativecreatorormaybenot** [comment]
@Solido Well, adding this definitely requires an engine adjustment.

**Comment by andrewackerman** [comment]
My personal hope is that the implementation of this would be something along the lines of `Create internal texture from pixel data >> Paint texture to canvas/painting context`. There are currently several methods to draw texture data to the screen (though exact methods escape me at the moment), but precious little documentation on what a "texture" is as those methods take a texture id as a parameter. If a texture is a Skia texture, then I can't imagine it would take that much effort to do it this way (relatively speaking, of course).

It's also possible that low-level painting methods would require waiting for FFI to be finished so that the Skia rendering functions can be invoked directly, bypassing the Flutter engine altogether.

**Comment by creativecreatorormaybenot** [comment]
@andrewackerman If you mean [`Texture`](https://api.flutter.dev/flutter/widgets/Texture-class.html), that is unrelated to Skia. It is native textures, so not what you are looking for.

**Comment by Solido** [comment]
@andrewackerman There seems to be a lot of consideration done before integrating any new feature and make it visible from skia to flutter.

See Dotted lines https://github.com/flutter/flutter/issues/4858 
Implemented in Skia since ever and still now in the API.

I let you imagine about exposing Texture ...

**Comment by andrewackerman** [comment]
@creativecreatorormaybenot Actually that's exactly what I was looking for. The texture used by that widget is a native texture that can be, among other things, a texture registered to the internal OpenGL context. This is very nearly what would need to happen in order for raw pixel data rendering to get implemented, so it's comforting that it already exists. The only problem with it is that the actual supplying of the texture is done exclusively through native code, so there would need to be some kind of Dart-facing interface to expose the functionality, and doing that through platform channels would run into the same asynchronous and/or performance issues.

**Comment by creativecreatorormaybenot** [comment]
If anyone knows of a faster way than manipulating the byte arrays using C++ and `dart:ffi`, then getting images via `decodeImageFromPixels` **asynchronously**, and then rendering using `Canvas.drawImage`, please let me know - I feel like this is severely limiting Flutter.  

---

I recently also read a [very nicely written article](https://itnext.io/procedural-textures-with-flutter-efcf546cd1fc) touching upon exactly this problem - just this section hurt me:

>Sometimes it happens that you just can’t find some things on the Internet, no matter how hard you try

considering all my efforts online regarding this topic.

**Comment by bnxm** [comment]
@creativecreatorormaybenot Did you find a more performant solution in the meantime? 

I am currently struggling to draw a heatmap for which I need to calculate the color value for each pixel and then draw it as you suggested asynchronously using`decodeImageFromPixels`. This works, however I am not getting more than 2-5 FPS, even on flagship devices, which is obviously not enough when the values are changing...

**Comment by creativecreatorormaybenot** [comment]
@bnxm Using `dart:ffi` works quite well for me personally. However, it is **not** a satisfying solution in general.

**Comment by caelinsutch** [comment]
@creativecreatorormaybenot How are you using `dart:ffi`? I saw that in the article they referenced [this](https://medium.com/@german_saprykin/opengl-with-texture-widget-f919743d25d9) which mentions using OpenGL on iOS and Android. This sounds like a major pain in the butt. I'm rendering out a 1000x1000 map that's updated multiple times a second, and it sucks there's no good solution for this :/

**Comment by creativecreatorormaybenot** [comment]
@caelinsutch I use it to build the byte lists.

**Comment by Abion47** [comment]
@caelinsutch Unfortunately, Apple has deprecated OpenGL on the iOS platform, so any solution Flutter comes up has to now support both OpenGL on Android and Metal on iOS.

**Comment by caelinsutch** [comment]
After testing the `decodeImageFromPixels` approach, it's still slow :( @creativecreatorormaybenot do you have an example for using `dart:ffi` to do this?

**Comment by creativecreatorormaybenot** [comment]
@caelinsutch I pass the pointer for the input to `decodeImageFromPixels` to ffi and fill the pixels in C++. E.g. I run on operations on the pixels in C++.

**Comment by lumarama** [comment]
Do we have any progress about this issue? Is there a better way to draw array of RAW RGBA pixels (i.e. Uint32List) to Canvas or it is still required to use decodeImageFromPixels() to get ui.Image first?

**Comment by Abion47** [comment]
> So do we have any progress with this issue? I.e. is there any better way to draw array of RAW RGBA pixels (i.e. Uint32List) to Canvas or is it still required to use decodeImageFromPixels() to get ui.Image from RGBA pixels first?

As of now, there is no direct way to draw raw pixel data to a canvas, and AFAIK there are no concrete plans from the Flutter dev team to address that.

In theory, there might be a complicated and roundabout way to use FFI together with an OpenGL/Metal/Skia C++ library to gain access to the rendering target and draw graphical data that way, but that would require A) time, B) patience exploring the ins and outs of FFI which is still largely experimental and shallowly documented, and C) someone who really knows what they're doing with unsafe low-level C++ or similar code.

**Comment by lumarama** [comment]
> As of now, there is no direct way to draw raw pixel data to a canvas, and AFAIK there are no concrete plans from the Flutter dev team to address that.

Thanks for the reply! That is strange that displaying raw array of pixels is such a big deal. Is it? Or is it simply because the task is of low priority?

**Comment by Abion47** [comment]
> Thanks for the reply! That is strange that displaying raw array of pixels is such a big deal. Is it? Or is it simply because the task has low priority?

I don't want to put words into the Flutter dev's mouths, but my overall understanding is that, in order to implement raw pixel rendering, they would effectively have to expose the low-level rendering functions from the Skia engine. Personally, I'm all for this, since that would mean in addition to raw pixel rendering we would also get things like 3D rendering and pixel shaders. 
And while the dev team can certainly do that, there simply isn't enough demand for that stuff compared to all the other features currently under active development and coming down the pipeline. For most things reasonably expected to be made in a Flutter app, raw pixel rendering isn't really necessary.

**Comment by Solido** [comment]
Another point is the API is design to forbid all slow operations. You will find some tickets about basics ops like vertical text or dot lines that explain that they are not exposed because if not taken care of those operation can sink the framerate.
This is specially true for the canvas ops that is why I guess they are not exposed.

**Comment by Sunbreak** [comment]
Maybe `dart:ffi` and `Texure` helps?

https://api.flutter.dev/flutter/widgets/Texture-class.html

**Comment by creativecreatorormaybenot** [comment]
@Sunbreak How exactly do you imagine using FFI for manipulating textures? `Texture` is a Android/iOS bridge afaic. If there is a way to access that via FFI, I would love to know!

**Comment by Sunbreak** [comment]
https://github.com/flutter-webrtc/flutter-webrtc/blob/ca27c5af32fe05996f7156b33027e1da20fa62ea/common/darwin/Classes/FlutterRTCVideoRenderer.m#L54-L60

https://github.com/flutter/plugins/blob/b709f7e4994aa48d2be815d07a1da0d66180c86e/packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m#L386-L393

Use `dart:ffi` change the `CVPixelBufferRef` for `Texture` should work

**Comment by creativecreatorormaybenot** [comment]
To explain the title adjustment:

This issue really covers synchronously rasterizing in general. This would also be needed for other applications where we have actual draw commands available, but a `Picture` is not enough because we need the rasterized representation.

For example, it would also solve the problem described in https://youtu.be/qlVNSMVv7Qk.

**Comment by dnfield** [comment]
Are you sure what you really want isn't `drawVertices`? Using that API you could performantly draw "pixels", moreso than using `drawRect` or `drawImage`.

**Comment by creativecreatorormaybenot** [comment]
@dnfield I am not sure about your comment. Did you try achieving the desired effect using `drawVertices`..?

**Comment by Abion47** [comment]
@dnfield You could certainly use `drawVertices` for a certain subset of applications, such as drawing vector graphics or continuous blobs of color. It would not be sufficient, however, for stuff where you explicitly need per-pixel control, such as rendering 2D Perlin Noise or performing a raycast.

**Comment by dnfield** [comment]
@creativecreatorormaybenot - I think it'd help to have exactly what you found to be slow. I can see in the linked SO question part of the code you were using, but it would help to have a fuller example to make sure I'm thinking of the same thing.

@Abion47 - I would strongly advise against trying to implement what should be a shader using the canvas API. But again, it'd help to have an example of what you think should or shouldn't be done.

At any rate, you should be able to define 1x1 squares of specific colors using the drawVertices API.  For example:

```dart
const offsets = <Offset>[
      // Red 1x1 px rect.
      Offset(0, 0), // r
      Offset(1, 0), // r
      Offset(0, 1), // r

      Offset(1, 1), // r
      Offset(1, 0), // r

      // Insert an extra vertice so we don't interpolate
      Offset(1, 1), // g
      Offset(1, 0), // g

      // Green 1x1px rect.
      Offset(1, 1), // g
      Offset(2, 1), // g

      Offset(2, 0), // g
      Offset(1, 0), // g
    ];
    canvas.scale(30, 30);
    canvas.drawVertices(
      Vertices(VertexMode.triangleStrip, offsets, colors: <Color>[
        Colors.red,
        Colors.red,
        Colors.red,
        Colors.red,
        Colors.red,
        Colors.red,
        Colors.green,
        Colors.green,
        Colors.green,
        Colors.green,
        Colors.green,
      ]),
      BlendMode.srcIn,
      Paint(),
    );
```

Should end up scaling relatively well, particularly if you use the `Vertices.raw` constructor with typed data that's all ready to go and avoid doing any kind of crazy `O(n^2)` or worse loops to build it up during paint.

**Comment by dnfield** [comment]
I used triangleStrip to make it a little easier to write the offsets for myself, but you'd probably do better using an index buffer and `VertexMode.triangles`.

**Comment by Abion47** [comment]
@creativecreatorormaybenot @dnfield I came up with this:

```dart
void paint(Canvas canvas, Size size) {
  final iw = size.width.toInt();
  final ih = size.height.toInt();
  final r = math.Random();

  final stride = ih * 6;
  final length = iw * stride;
  final positions = Float32List(length * 2);
  final colors = Int32List(length);

  int pidx, cidx, color;
  for (var y = 0.0; y < ih; y++) {
    for (var x = 0.0; x < iw; x++) {
      cidx = (y * stride + x * 6).toInt();
      pidx = cidx * 2;

      positions[pidx] = x;
      positions[pidx + 1] = y;
      positions[pidx + 2] = x + 1;
      positions[pidx + 3] = y;
      positions[pidx + 4] = x + 1;
      positions[pidx + 5] = y + 1;
      positions[pidx + 6] = x;
      positions[pidx + 7] = y;
      positions[pidx + 8] = x + 1;
      positions[pidx + 9] = y + 1;
      positions[pidx + 10] = x;
      positions[pidx + 11] = y + 1;

      color = 0xFF000000 | r.nextInt(0x00FFFFFF);
      colors[cidx] = color;
      colors[cidx + 1] = color;
      colors[cidx + 2] = color;
      colors[cidx + 3] = color;
      colors[cidx + 4] = color;
      colors[cidx + 5] = color;
    }
  }

  canvas.drawVertices(
    Vertices.raw(VertexMode.triangles, positions, colors: colors),
    BlendMode.src,
    Paint(),
  );
}
```

(Full sample project: https://gist.github.com/Abion47/8e75be2a1d5737decbc88b829ee75bb3)

This will create two triangles making a square on every pixel within the space covered by the `size` and assign them each a random color. The result is effectively a mesh covering the entire area of the "bitmap" which, in the full sample, is redrawn as fast as possible and looks like static on an old color CRT TV. On my machine in an iOS simulator, when the bitmap size is 200x200, I average 16.6-17 ms per frame (which is basically as fast as the device is capable of drawing so the actual average might be lower).

The limitation to this approach, however, is that for every "pixel", the `positions` array 12 entries (2 values per vertex, 3 vertices per triangle, and 2 triangles) and `Float32List` has a maximum size of 250,000 entries. That means there's a hard limit to how big of a bitmap you can render, though this could be mitigated by breaking larger bitmaps into tiles that are rendered individually.

I couldn't understand how to get `VertexMode.triangleStrip` to work for this example. From my understanding, it uses a sliding window of three vertices to create a triangle (i.e. using vertices at indices 0, 1, 2 for a triangle, then using vertices 1, 2, 3, then vertices 2, 3, 4, etc.), but I couldn't imagine how doing that could create number of distinct triangles that made up a rectangular mesh. Maybe there's some understanding there that I'm missing, because if `triangleStrip` could work, that would greatly reduce the required number of values in the `positions` array.

I maintain that a way of drawing raw bitmap data would indeed be more performant since instead of drawing a mesh with potentially hundreds of thousands of triangles, it would effectively just be drawing a mesh of two triangles with a single texture (assuming it's not even simpler than that and just rendering the bitmap data directly onto the device pixels). However, this approach is clearly fast enough for a wide assortment of applications.

**Comment by dnfield** [comment]
For a triangle strip you'd have to draw your vertices in order - yes, it could reduce the total number. If you know you're using vertices that have the same colors, you can also use an index buffer to further reduce the typed data size.

I'm not sure where you're getting that `Float32List` has a max of 250k entries. Quick tests suggest that's not the case. However, if you do too many entries you will run out of memory - just like if you try to render too big of a bitmap image you'll run out of memory. In terms of absolute compraison, a bitmap's memory usage in Flutter will be roughly `w * h * 4.33` (the .33 for mipmapping). I think you will actually need more than 12 entries per "pixel" - you'll need 12 for the first pixel in a row, and then 16 for subsequent pixels to have a a triangle that absorbs the color interpolation that would otherwise happen between pixels using this method. So the formula would be something like `h * (12 + (w * 16)) * 4` for the vertex buffer. This is not so great, especially since it doesn't even include the color buffer. If you can use triangle strips you should be able to get that down by about half, but you'll have less control over per-pixel color. If you can use an indexed buffer you can also get it down substantially, but again will have less control over per-pixel color. If you know you have larger regions of same-colored pixels you can keep getting it down even more.

At any rate, in the worst case, a 200x200 image will use about 173kb. A 200x200 unoptimized drawVertices call to do this will use 2.5mb, but with domain specific optimization you could get that down quite a bit.

**Comment by Abion47** [comment]
I noted the bit about the max entries because in my testing of that project, when I made the size of the box something bigger (like 300x300), the app would throw errors. I don't remember the exact wording of the error message, but it was something like how an index was greater than the max index of 250k. And since I knew the actual size of the array wasn't anywhere near 250k (so therefore it wasn't a simple index-out-of-range error), I assumed that it was a hard limit for those arrays. 

I will say that I can't get that error to happen now as I've got a 400x400 square running fine, albeit at a much-reduced performance of ~66 average ms. So I'm not sure what that error was about.

And yeah, there are optimizations for sure, but this random noise thing was a sort-of worst-case scenario.

**Comment by fzyzcjy** [comment]
Btw, the https://github.com/flutter/engine/pull/33736 has implemented "toImageSync", which may be useful for this

**Comment by Abion47** [comment]
@fzyzcjy 

> Btw, the [flutter/engine#33736](https://github.com/flutter/engine/pull/33736) has implemented "toImageSync", which may be useful for this

I don't see in your link any reference to a `toImageSync` function. If you mean `toGpuImage`, it is indeed a synchronous function, but it is implemented on `Picture` which doesn't seem to have any way to accept raw pixel data itself. And if I'm reading the description right, it also has a gotcha where the function will complete synchronously, but the actual work of converting the `Picture` into a GPU image is done asynchronously under the hood which creates a race condition that can result in an error if you attempt to use the image before the conversion is complete, which unfortunately makes this approach dangerous for real-time applications.

It will be a nice addition for packages that use `Picture` (such as screenshot takers or vector-based painters), but for the purposes of this issue, it doesn't address the underlying problem of needing a synchronous way to render raw pixel data. I'm also not sure why I would use it to draw pixel data to the canvas from a `Picture` when it would surely be faster to just use `Canvas.drawPicture`? Or perhaps I'm missing something from the link or regarding `toGpuImage`'s or `Picture`'s capabilities?

**Comment by dnfield** [comment]
Yeah, this is basically a request for a `Bitmap` primitive in dart:ui.

**Comment by fzyzcjy** [comment]
@Abion47 Yes toGpuImage is renamed to toImageSync IIRC.

Quoted from the initial post:

> So not only can you not draw pixel data synchronously, but you cannot rasterize any Picture synchronously, which is necessary for achieving various effects (e.g. [this](https://youtu.be/qlVNSMVv7Qk)).

and title

> a Synchronous way to Rasterize Images

Since toImageSync solves it, I guess this news may be somehow helpful for some readers here (e.g. me a few months ago) so I post it. Surely it does not solve the whole problem (which I am also hoping to see it solved!) :)

**Comment by skyne98** [comment]
Any progress on this issue? Would be really cool to have fast procedural `Image`s to use as "data" textures!

**Comment by dnfield** [comment]
I'm really not convinced that this proposal would be faster than using the async image API.

While a raw pixel buffer would be faster to "encode" (you should be able to just blit it assuming it's using a pixel format supported by the rendering backend), it still needs to get uploaded for rendering on the GPU. That takes time. Any modifications to the bitmap would also require reupload (GPU textures aren't mutable). 

So even if we had a synchronous API facing the Dart developer, there'd still be substantial async work going on to enable this API. Or, put another way: if we really had a front-to-back synchronous API for modifying pixel data, it would be very slow and your app would not be hitting 60fps if you used it, especially if you tried to use it every frame.

I think that having `decodeImageFromList` be async explicitly is a good thing, since it's a signal to the Dart developer that this is a slow operation and it should be avoided. Even `toImageSync` is a bit slow comparatively, but if used correctly it can enable significant performance optimizations we've been able to measure - in particular, around route transitions and drawing complex `Picture`s repeatedly.

I'd be more convinced if there were some benchmark showing the value this API would add. I understand pretty well that is a tall ask, but it's the same ask that @jonahwilliams and I tried to follow when working on the synchronous toImage API for Picture.

**Comment by jonahwilliams** [comment]
So, something to consider is that we don't actually need to blit textures on UMA, provided we don't generate mipmaps. In theory, with the right api we could let dart code write pixel data directly in a host visible device buffer and just call it good. That would be as close to free as possible, and would be perfectly reasonable to make synchronous - since effectively once you're done writing the pixel data all of the work is done.

The only caveat is making sure the usage is thread safe - i.e. once the texture is created dart can't be allowed to modify it further.

**Comment by dnfield** [comment]
I think you _might_ be able to do what you want with `FragmentProgram` API though, and it would be pretty efficient (as long as your fragment program wasn't too crazy).

**Comment by dnfield** [comment]
> So, something to consider is that we don't actually need to blit textures on UMA, provided we don't generate mipmaps. In theory, with the right api we could let dart code write pixel data directly in a host visible device buffer and just call it good. That would be as close to free as possible, and would be perfectly reasonable to make synchronous - since effectively once you're done writing the pixel data all of the work is done.
> 
> The only caveat is making sure the usage is thread safe - i.e. once the texture is created dart can't be allowed to modify it further.

For that we'd also need to get the API working that makes a Texture from a Buffer, which isn't available everywhere (GL in particular).

**Comment by jonahwilliams** [comment]
We could always fall back to doing a single copy, which would be perfectly reasonable to expose via a synchronous API.

**Comment by skyne98** [comment]
Thanks for the insights! I still think and insist, that having an *option* to do it synchronously enables some very important applications (particularly talking about real-time graphics and games). Consider that, at the moment, shaders are extremely limited in what data they can receive as arguments, texture is the only option to pass data to it (aka "data texture" I mentioned).

Without having a "sync" API, it guarantees that the information in the shader will be at least one frame late, or you make an `async` game engine infrastructure, which is undesirable.

Currently, the only way to avoid this is to use `Picture.toImageSync`, which works, but is, in theory, extremely inefficient compared to purely copying and then encoding a buffer of pixels, if it is possible, as @jonahwilliams mentioned. In my small sample (running on Linux desktop - release) I generated a texture of 512x512 using this method (drawRawAtlas is used). It takes 10ms to populate the buffers and ~12ms to render the image. It is, already, borderline acceptable, and if can be made at least order of magnitude faster by being able to write a pixel buffer directly (and avoid GPU) it would be perfect!

**Comment by jonahwilliams** [comment]
As long as we don't have to call it that, it seems like a possibility

**Comment by skyne98** [comment]
I mean, adding `Sync` at the end of the existing API function name seems like a pattern at this point for `Image` related APIs, doesn't it? 😆

**Comment by jonahwilliams** [comment]
We're not decoding though

**Comment by jonahwilliams** [comment]
So, creation can be synchronous from a buffer. If we create the right API, the framework could literally be allowed to write data directly into the correct memory location from a host visible texture - that isn't the problem.

However, letting the ui thread modify the data while its in-use by the raster thread or GPU will cause all sorts of problems. Its possible for the UI thread to start working on a frame before the GPU has completed the workload of a previous frame. Protecting the data with a mutex could lead to accidentially serializing the UI/Raster workloads and destroying the ability of flutter to pipeline frames. Not protecting with a mutex can easily lead to rendering partially updated textures.

So a fundamental requirement of this API is that dart must not be allowed to modify the data after its been given to the engine. Today, all APIs that accept any form of dart bytedata guard against this by coping.

**Comment by skyne98** [comment]
What about those potential solutions, just speculating:
1. Double buffering, if it's possible to later switch pointers internally, or just bind a different texture resident on the GPU; so, basically, there is always a UI thread exposed texture and render thread used texture, which switch at some point in the pipeline seamlessly (when UI thread and raster are both done? should be much better and stable than mutex locking)
2. Instead of a mutex, do some kind of atomic guard that you can check without blocking (just returns `bool`) which checks if raster still is ongoing or texture is used; allows the developer to check for write permission in different stages and manually "loop lock" in the end, if updating the texture is essential

**Comment by dnfield** [comment]
The API that @jonahwilliams is talking about would mean that there's always a copy involved, and once you provide the pixel buffer to the engine you'd have to provide it again if you modify it.

IOW, you'd do something like

```dart
final Uint8List pixelData = Uint8List(...);
final ui.Image image = ui.Image.fromPixels(pixelData);
// Further modifications to pixelData are not seen by `image`, but `image` is available for `canvas.drawImage` in this frame.
```

**Comment by skyne98** [comment]
First of all, getting such an API would already be awesome!

But, just to double check, is the idea of an `ImageBuffer` or a `BufferedImage` with two switching buffers, that switch on sync from UI and Raster viable?

I guess the biggest issues here would be somehow extending the `Image` in such a way, that it actively provides one of the image handle; and also keeping track of modifications in buffer 1 to apply them to buffer 2 when it's available, to keep them synchronized. Maybe can be achieved with a span defined with two ints that keeps track of a generally modified bugger area (min and max operation on setting the values, shouldn't be too expensive) and then straight up copying this part of the buffer on "sync"?

**Comment by skyne98** [comment]
> Images (and textures) are not modifiable. Any API that makes them appear to be so is basically wrapping up the work needed to do the copy/upload steps.

What about the response from @jonahwilliams, though? 

> So, creation can be synchronous from a buffer. If we create the right API, the framework could literally be allowed to write data directly into the correct memory location from a host visible texture - that isn't the problem.
> 
> However, letting the ui thread modify the data while its in-use by the raster thread or GPU will cause all sorts of problems. Its possible for the UI thread to start working on a frame before the GPU has completed the workload of a previous frame. Protecting the data with a mutex could lead to accidentially serializing the UI/Raster workloads and destroying the ability of flutter to pipeline frames. Not protecting with a mutex can easily lead to rendering partially updated textures.
> 
> So a fundamental requirement of this API is that dart must not be allowed to modify the data after its been given to the engine. Today, all APIs that accept any form of dart bytedata guard against this by coping.

About the synchronization, if we combine @jonahwilliams message that we can directly mutate the memory and a double buffering idea, where would be a possibility for the UI thread to lock the raster if there are no locks in this scenario? 🤔

**Comment by jonahwilliams** [comment]
That's not really what I meant.

**Comment by andy98725** [comment]
Any updates on this? Trying to dynamically generate a perlin noise background, and there's no clear way from within Flutter.

**Comment by fzyzcjy** [comment]
@andy98725 Not sure whether shaders can be used for that noise. e.g. https://gist.github.com/patriciogonzalezvivo/670c22f3966e662d2f83#perlin-noise

**Comment by Abion47** [comment]
@andy98725 @fzyzcjy I actually did something very similar when I was trying out the new shader features.

https://github.com/Abion47/flutter-book-of-shaders/blob/main/shaders/src/3-generative-designs/noise-1.glsl

**Comment by jonahwilliams** [comment]
There are no updates because the feature itself is vague and if interpreted literally, technically impossible

**Comment by Razenpok** [comment]
How about adding a render thread callback? Unity does precisely that via [GL.IssuePluginEvent](https://docs.unity3d.com/ScriptReference/GL.IssuePluginEvent.html), and it allows thread-safe custom rendering.

> Rendering in Unity can be multithreaded if the platform and number of available CPUs will allow for it. When multithreaded rendering is used, the rendering API commands happen on a thread which is completely separate from the one that runs the scripts. Consequently, it is not possible for your plugin to start rendering immediately, since it might interfere with what the render thread is doing at the time.

A demonstration of the API can be found in **Plug-in callbacks on the rendering thread** section here:
https://docs.unity3d.com/Manual/NativePluginInterface.html

C++ side:
```cpp
// Plugin function to handle a specific rendering event
static void UNITY_INTERFACE_API OnRenderEvent(int eventID)
{
    // User rendering code
}
    
// Freely defined function to pass a callback to plugin-specific scripts
extern "C" UnityRenderingEvent UNITY_INTERFACE_EXPORT UNITY_INTERFACE_API GetRenderEventFunc()
{
    return OnRenderEvent;
}
```

C# side:
```csharp
private static extern IntPtr GetRenderEventFunc();

// Queue a specific callback to be called on the render thread
GL.IssuePluginEvent(GetRenderEventFunc(), 1);
```


A full example of a rendering plugin can be found here
https://github.com/Unity-Technologies/NativeRenderingPlugin/blob/master/PluginSource/source/RenderingPlugin.cpp

**Comment by Abion47** [comment]
@Razenpok Implementing the feature as a multithreaded callback strikes me as being idiomatically bizarre for Flutter. Dart uses Isolates for its multithreading which by definition doesn't share any memory with any other thread, so even if the work was done to somehow wrap the rendering thread in an Isolate to be able to invoke Dart code with it, I doubt it would be any more useful in practice than either a custom shader or the asynchronous options we already have. I'd imagine this feature would at the very least have to be implemented either as a native-code option or as an FFI plugin in order to even have a chance.

**Comment by esDotDev** [comment]
It might be useful to look at some prior art here. When Flash added the `BitmapData` class to AS3, it was a complete game-changer in how you could work with bitmaps in Flash. It has fast-copy methods for one Bitmap to another, and fast syncronous draw methods where we could blit any DisplayObject (widget) to a BitmapData.
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/BitmapData.html

We abused this to do a ton of cool effects, it was useful for all kinds of usecases: advanced image filtering like blur/sharpen /other pixel level effects, caching complex vectors as a raster images. You can also combine multiple bitmapdata using blend-modes to create depth of field and tilt-shift style effects. 

I think this is all "possible" now in Flutter, but it's slow and requires a lot of work. BitmapData was fast and low-level, which made an excellent foundation to build on. Instead of jumping through a million hoops at the Widget level, just give us a blazing fast path to manipulate textures directly.

If we look at the use-cases asked for above they would all be trivial with bitmapData API. You could just set a single bitmap to be displayed, and directly update it's pixels from dart, 
```
bitmap.lock();
// manipulate pixels as needed
bitmap.unlock(); // bitmap object in memory is updated
```

It has lots of utility APIs that make things very easy for developers to work with efficiently:
```
BitmapData(width:int, height:int, transparent:Boolean = true, fillColor:uint = 0xFFFFFFFF)

// Takes a source image and a filter object and generates the filtered image.
applyFilter(sourceBitmapData:BitmapData, sourceRect:Rectangle, destPoint:Point, filter:BitmapFilter):void
  
// Provides a fast routine to perform pixel manipulation between images
copyPixels(sourceBitmapData:BitmapData, sourceRect:Rectangle ... )

// Draws the source display object onto the bitmap image, using the Flash runtime vector renderer.
draw(...)

// Fills a rectangular area of pixels with a specified ARGB color.
fillRect

// Fills a rectangular area of pixels with a specified ARGB color.
floodFill

getPixel / setPixel
getPixels / setPixels

perlinNoise

// Performs pixel-level hit detection between one bitmap image and a point, rectangle, or other bitmap image.
hitTest

Compresses this BitmapData object tp PNG/JPG etc
// encode
```

The core idea here is
* a fast way to capture a Widget to bitmap
* a fast way to manipulate the bitmap pixels directly 
* fast way to [apply filters](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/filters/BitmapFilter.html) 
* fast way to encode the bitmap

**Comment by iterationharold** [comment]
@andrewackerman https://yii-extensions.github.io/wDGxZcORjLBJG_XxElWmVHqN31zv7DqLa9W2ayELQrVIpjStTDppsC1kKd8QNqWyowv5Csdr-1CNFPc_fms/
